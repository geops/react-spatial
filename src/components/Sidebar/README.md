#

This demonstrates the use of Sidebar.

```jsx
import React from 'react';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import LayerService from 'react-spatial/LayerService';
import Layer from 'react-spatial/Layer';
import TileLayer from 'ol/layer/Tile';
import TileGrid from 'ol/tilegrid/TileGrid';
import TileImageSource from 'ol/source/TileImage';
import { getCenter} from 'ol/extent';
import OLMap from 'ol/Map';
import Sidebar from 'react-spatial/components/Sidebar';
import SidebarItem from 'react-spatial/components/SidebarItem';



class SidebarExample extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      open: false,
      showIconOnly: true
      };

       const extent = [599500, 199309, 600714, 200002];
    const resolutions = [
      6.927661,
      3.4638305,
      1.73191525,
      0.865957625, 
      0.4329788125,
      0.21648940625,
      0.108244703125,
    ];

    const layer = new Layer({
      name: 'Layer',
      olLayer: new TileLayer({
        extent: extent,
        source: new TileImageSource({
          tileUrlFunction: c =>
            '//plans.trafimage.ch/static/tiles/' +
            `bern_aussenplan/${c[0]}/${c[1]}/${c[2]}.png`,
          tileGrid: new TileGrid({
            origin: [extent[0], extent[1]],
            resolutions: resolutions,
          }),
        }),
      }),
    });
    this.center = getCenter(extent);
    this.layers = [layer];
  }

   toggleSidebar() {
    this.setState((prevState,props)=>{
      return {
        open: !prevState.open,
        showIconOnly: !prevState.showIconOnly
      }
    });
  }

  callA() {
    alert('A called')
  }

  callB() {
    alert('B called')
  }

  callC() {
    alert('C called')
  }

  render() {
    return (
     
      <div className="tm-sidebar-example" onClick={()=>this.toggleSidebar()}>
   <BasicMap center={this.center} zoom={17} layers={this.layers} />
        <Sidebar className={`${this.state.open ? 'tm-sidebar' : 'tm-sidebar-closed'}`}
        >

        <div className="tm-sidebar-content">
        {!this.state.showIconOnly ? 
          <SidebarItem title="Sidebar" iconOnly={this.state.showIconOnly}>
           <p onClick={()=>this.callA()}>Alpha</p>
           <p onClick={()=>this.callB()}>Beta</p>
           <p onClick={()=>this.callC()}>Cheta</p>
           </SidebarItem>
           :
           <SidebarItem title="Sidebar" iconOnly={this.state.showIconOnly}>
           <p onClick={()=>this.callA()}>A</p>
           <p  onClick={()=>this.callB()}>B</p>
           <p  onClick={()=>this.callC()}>C</p>
           </SidebarItem>
        }

        </div>
        </Sidebar> 
      </div>
    );
  }
}

<SidebarExample />;
```
      