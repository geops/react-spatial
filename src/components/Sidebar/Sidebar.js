import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Button from '../Button';

const propTypes = {
  /** Open or close the  */
  open: PropTypes.bool,
  title: PropTypes.string,
  position: PropTypes.oneOf(['left', 'right']),
  modal: PropTypes.bool,
  className: PropTypes.string,
  children: PropTypes.node,
  onModalClick: PropTypes.func,
  collapseWidth: PropTypes.number,
};

const defaultProps = {
  open: false,
  title: null,
  position: 'left',
  modal: true,
  className: 'tm-sidebar',
  children: null,
  onModalClick: () => {},
  collapseWidth: 0,
};
let timeout;

const Sidebar = ({
  open,
  title,
  position,
  modal,
  className,
  children,
  onModalClick,
  collapseWidth,
}) => {
  const [shouldRemoveElt, setRemoveElt] = useState(!open);

  // We add the SideBar element before opening it.
  // We remove the SideBar element after closing it.
  useEffect(() => {
    window.clearTimeout(timeout);
    timeout = window.setTimeout(
      () => {
        setRemoveElt(!open);
      },
      open ? 5 : 300,
    );
  });

  const isClosed = !open && shouldRemoveElt;

  // If the sidebar is closed and collapseWidth is 0, we remove
  // completely the SideBar from the tree.
  if (isClosed && collapseWidth === 0) {
    return null;
  }

  const isClosing = open && !shouldRemoveElt;
  const classToggle = isClosing ? 'open' : 'close';
  const classMain = `${className} ${className}-${position} ${className}-${classToggle}`;
  const classModal = `${className}-modal ${className}-modal-${classToggle}`;
  const style = !open ? { width: collapseWidth } : undefined;
  const header = title ? <header>{title}</header> : null;

  // If the sidebar is closed, we remove the modal background.
  const modalBg =
    !modal || isClosed ? null : (
      <Button className={classModal} onClick={onModalClick} />
    );

  return (
    <>
      {modalBg}
      <div className={classMain} style={style}>
        {header}
        {children}
      </div>
    </>
  );
};

Sidebar.propTypes = propTypes;
Sidebar.defaultProps = defaultProps;

export default Sidebar;
