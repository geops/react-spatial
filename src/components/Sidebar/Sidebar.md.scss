 .tm-sidebar-example {
  position: relative;

.tm-sidebar {
  position: absolute;
  left: 0;
  top: 55px;
  bottom: 0;
  width: 12%;
  max-width: 120px;
  background: grey;
  z-index: 200;
  box-shadow: 0 3px 7px rgba(0,0,0,0.5);
  transform: translateX(-100vw);
  transition: transform 0.4s ease-in;
  overflow: hidden;
  align-items: center;
}

.tm-sidebar.Open {
  transform: translateX(0);
}

.tm-sidebar ul {
  list-style: none;
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100%;
}

.tm-sidebar li {
  margin: 0.5rem 0;
  color: white;
}

.tm-sidebar li:active,
.tm-sidebar li:hover {
    color: #fa923f;
}

.headerClass {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  background-color: #521751;
  height: 56px;
}

.header__navigation {
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 1rem;
}

.header_navigation__items ul{
  list-style-type: none;
  margin: 0;
  padding: 0;
  display: flex;
}

.header_navigation__items li {
  padding: 0 0.5rem;
  display: inline;
  color: white;
}

.spacer {
  flex: 1;
}

.toggle-button {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 24px;
  width: 36px;
  background: transparent;
  border: none;
  padding: 0;
  box-sizing: border-box;
}

.toggle-button:focus {
  outline: none;
}

.toggle-button__line {
  width: 30px;
  height: 2px;
  background: white;
}

@media (max-width: 768px) {
  .header_navigation__items {
      display: none;
  }
}

@media (min-width: 769px) {
  .toggle-button {
      display: none;
  }

  .tm-sidebar {
    display: none;
  }
}

}
