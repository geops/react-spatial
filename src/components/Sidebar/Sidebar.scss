.tm-sidebar {
  position: absolute;
  top: 0;
  background-color: white;
  height: 100%;
  transition: width 300ms;
  width: 0;
  overflow-x: hidden;
}

.tm-sidebar-open {
  width: $sidebar-open-width;
}

.tm-sidebar-left {
  left: 0;
  border-right: 2px solid #eee;
}

.tm-sidebar-right {
  right: 0;
  border-left: 2px solid #eee;
}

.tm-sidebar-modal {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: $gray-light;
  opacity: 0;
  transition: opacity 300ms;
}

.tm-sidebar-modal-open {
  opacity: 0.6;
}
