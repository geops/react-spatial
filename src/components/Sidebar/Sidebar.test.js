import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Sidebar from './Sidebar';

configure({ adapter: new Adapter() });

describe('Sidebar', () => {
  describe('should match snapshot.', () => {
    test('when closed.', () => {
      const component = renderer.create(<Sidebar />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('when opened', () => {
      const component = renderer.create(<Sidebar open />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
