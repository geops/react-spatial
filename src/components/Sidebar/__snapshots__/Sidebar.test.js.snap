// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should match snapshot. when closed. 1`] = `null`;

exports[`Sidebar should match snapshot. when opened 1`] = `
Array [
  <div
    className="tm-sidebar-modal tm-sidebar-modal-open"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  />,
  <div
    className="tm-sidebar tm-sidebar-left tm-sidebar-open"
  />,
]
`;
